core/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -124,7 +124,13 @@ public Weight createWeight(Query query, boolean needsScores) throws IOException
                 profile.stopAndRecordTime();
                 profiler.pollLastElement();
             }
-            return new ProfileWeight(query, weight, profile);
+
+            if (weight instanceof ProfileWeight) {
```

Comment:
I think this needs to happen before any profiling is done.  The profiler is a bit wonky in how it works, due to the way Lucene works: when query components are timed (via `startTime`) they are pushed onto a stack.  Lucene queries are hierarchical in nature, so calling `createWeight()` on one query will invoke `createWeight()` on any child queries.  The stack tracks the order and relationship between query components.

Once the method call returns, the time is recorded and the component popped off the stack.  After everything is done, the profiler uses the information gained during the stack push/pop to rebuild the tree structure of timnigs.

So, if a normal weight is returned instead of a `ProfileWeight`, the accounting at the end will likely get messed up because `n` times were recorded, but `n-1` weights were wrapped.

I believe if you move the `return weight` up before the `startTime()` it'll behave better.